### PR TITLE
Document internal API authentication

### LINKED ISSUE #389: Document Easy Auth authentication story
We've established the use of App Service authentication (aka "Easy Auth") as a mechanism for allowing our sub-systems to utilize our internal APIs. Microsoft documentation around the process is broken up across many different topics and sections. Consequently, we do not have a concise and tailored story to use as reference.

Documentation should include:
- Definitions of the relevant Azure resources (app registration, service principal, managed identity, system identity, etc).
- High-level diagram(s) of authentication flow
- Clarity on the boundaries for Azure Resource Manager and Azure Active Directory

### PR DESCRIPTION
**Pull request by kingcomma**
Adds documentation on our process for securing internal APIs. Primarily intended for use by members of the security team and system developers, with the goal of giving a conceptual understanding. I've intentionally left out code-level detail which I'm thinking is better defined in future `Piipan.Shared` docs.

@ryanhofdotgov Especially interested if anything in here can be reworded to better align with common security nomenclature.

@echappen and @heymatthenry If you have time to review, especially interested in your thoughts as developers who haven't yet had to get too far into these weeds.

Closes #389.

### CODE REVIEW

**Comment by echappen** [comment]
@kingcomma wow this is a lot. Thanks for writing all this up! I'm sure I'll have more thoughts on its utility when I start implementing secure auth, but atm this looks great to me.

**Review by echappen — APPROVED** [review]

File: docs/securing-internal-apis.md [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+
+1. The client identifies itself to the server's authentication service and requests an access token
+1. The server's authentication service authenticates the client, performs authorization, and responds with an access token
+1. The client sends a standard HTTP request to the server, including the access token in the request's `Authorization` header
+1. The server's authentication service validates the access token and allows the request through to the server
+1. The server returns a response to the client
+
+## Resources involved
+
+In the Azure environment, the resources involved in authenticated client–server communication are divided between two domains: Azure resources (e.g., Azure Functions, App Service apps, database clusters, etc) and Azure AD resources (e.g., application objects, service principals, user accounts, etc). Azure resources provide the application logic (i.e., sending requests and responses) while their related Azure AD resources handle authentication and authorization.
```
Comment (ryanhofdotgov) [inline docs/securing-internal-apis.md:34]:
```suggestion
In the Azure environment, the resources involved in authenticated client–server communication are divided between two domains:
- Azure resources (e.g., Azure Functions, App Service apps, database clusters, etc) and 
- Azure AD resources (e.g., application objects, service principals, user accounts, etc).
Azure resources provide the application logic (i.e., sending requests and responses), while their related Azure AD resources handle authentication and authorization.
```

**Comment by ryanhofdotgov** [comment]
@kingcomma Fantastic write-up – I have a few minor suggestions that I'm writing up. Mind holding up on the merge?

File: docs/securing-internal-apis.md [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+
+1. The client, represented by its system-assigned identity (i.e., service principal), identifies itself to the server's local service principal and requests an access token by sending a request to the application object's `application ID URI`.
+1. The server's authentication service, Azure AD, authenticates the client, ensures it is assigned an application role, performs authorization, and responds with an authorized access token
+1. The client sends a standard HTTP request to the server and includes the access token in the request's `Authorization` header
+1. The server's authentication service, Azure AD, validates the access token and allows the request through to the server
+1. The server returns a response to the client
+
+An illustrated version of this flow:
+
+<p align="center">
```
Comment (ryanhofdotgov) [inline docs/securing-internal-apis.md:73]: Could a link to the upstream Google Doc of this diagram be documented on this PR, or perhaps in a hidden tag on in the HTML or something?

Not for this PR, but the bits regarding the AD side could benefit from some additional clarity; e.g., what does dotted line mean, what does solid line mean if they are used to connect service principal boxes, etc. That part is hard to follow in comparison to the network-flow-oriented part of the diagram.

**Review by ryanhofdotgov — COMMENTED** [review]
Super valuable doc! 🏆

Can be merged as-is, but I have some small suggested changes that I'll leave to you to apply or not.
